### PR TITLE
Move calls to bpLocation's unionmembers to main thread.

### PR DESCRIPTION
**Pull request by pieandcakes**
This should fix part of #350 's issues in BindAsync().
@gregg-miskelly @chuckries 

VSTS bug 451112

### CODE REVIEW

File: src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs [hunk 1]
```diff
@@ -248,29 +271,25 @@ int IDebugPendingBreakpoint2.Bind()
...
+                                    codeAddress = ((AD7MemoryAddress)_codePosition).Address;
                                     break;
                                 }
                             case enum_BP_LOCATION_TYPE.BPLT_CODE_FILE_LINE:
                                 {
-                                    IDebugDocumentPosition2 docPosition = HostMarshal.GetDocumentPositionForIntPtr(_bpRequestInfo.bpLocation.unionmember2);
-
                                     // Get the name of the document that the breakpoint was put in
-                                    EngineUtils.CheckOk(docPosition.GetFileName(out documentName));
+                                    EngineUtils.CheckOk(_docPosition.GetFileName(out documentName));
```
Comment (pieandcakes) [inline src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs]: @gregg-miskelly I guess my question is in these calls where its GetFileName or GetRange, I should be caching that in the bind call also?

**Comment by pieandcakes** [comment]
@dotnet-bot test please

**Comment by pieandcakes** [comment]
@dotnet-bot test this please

**Comment by pieandcakes** [comment]
@paulmaybee I don't know if changing the DebugAdapterHost API is going to affect you, but I want to make sure you know about it.

File: src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs [hunk 2]
```diff
@@ -116,14 +130,13 @@ public AD7DocumentContext GetDocumentContext(ulong address, string functionName)
         {
             if ((enum_BP_LOCATION_TYPE)_bpRequestInfo.bpLocation.bpLocationType == enum_BP_LOCATION_TYPE.BPLT_CODE_FILE_LINE)
             {
-                IDebugDocumentPosition2 docPosition = HostMarshal.GetDocumentPositionForIntPtr(_bpRequestInfo.bpLocation.unionmember2);
                 string documentName;
-                EngineUtils.CheckOk(docPosition.GetFileName(out documentName));
+                EngineUtils.CheckOk(_docPosition.GetFileName(out documentName));
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs]: Can we just use `_documentName`/`_startPosition`?

File: src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs [hunk 3]
```diff
@@ -34,7 +34,21 @@ internal class AD7PendingBreakpoint : IDebugPendingBreakpoint2
         private bool _deleted;
         private bool _pendingDelete;
 
-        public DebuggedProcess DebuggedProcess {  get { return _engine.DebuggedProcess;  } }
+        private IDebugFunctionPosition2 _functionPosition;
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs]: Can we avoid storing the AD7 objects (`_functionPosition`, `_codePosition`, `_docPosition`) as member variables if we don't use them in GetDocumentContext?

File: src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs (hunk 1 above)
Comment (gregg-miskelly) [reply src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs]: I think we should use the cached info. Really this whole thing is a bit wrong in that we should really be asking GDB where the break point bound. But if we aren't going to do that, it is better to use the info we cached then to re-ask VS for it.

File: src/DebugEngineHost.Stub/DebugEngineHost.ref.cs [hunk 4]
```diff
@@ -64,6 +64,15 @@ public static HostUIIdentifier GetHostUIIdentifier()
         {
             throw new NotImplementedException();
         }
+
+        /// <summary>
```
[the remark this answers is not in the document]
Comment (pieandcakes) [reply src/DebugEngineHost.Stub/DebugEngineHost.ref.cs]: @gregg-miskelly Are you saying move the method to the HostMarshal class itself or make the calls to this new API from all the methods in the VS implementation of the HostMarshal class?

[the remark this answers is not in the document]
Comment (paulmaybee) [reply src/DebugEngineHost.Stub/DebugEngineHost.ref.cs]: Agree with Gregg here. Move the OnMainThead checks into the VS implementation of HostMarshal.Xxx. Then no need to worry about breaking any other host implementations.

[the remark this answers is not in the document]
Comment (pieandcakes) [reply src/DebugEngineHost.Stub/DebugEngineHost.ref.cs]: @gregg-miskelly @paulmaybee these should be fixed now. Thanks

**Pull request merged**